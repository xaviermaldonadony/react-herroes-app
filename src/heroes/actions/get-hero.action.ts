import { heroApi } from '../api/hero.api';
import type { Hero } from '../types/hero.interface';

const BASE_URL = import.meta.env.VITE_API_URL;

export const getHeroAction = async (idSlug: string) => {
  const { data } = await heroApi.get<Hero>(`/${idSlug}`);

  return {
    ...data,
    image: `${BASE_URL}/images/${data.image}`,
  };
};
