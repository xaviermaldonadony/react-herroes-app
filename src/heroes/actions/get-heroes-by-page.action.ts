import { heroApi } from '../api/hero.api';
import type { HeroesResponse } from '../types/get-heroes.response';

const BASE_URL = import.meta.env.VITE_API_URL;

export const getHeroesByPageAction = async (
  page: number,
  limit: number = 6,
  category: string = 'all'
): Promise<HeroesResponse> => {
  page = isNaN(page) ? 1 : page;
  limit = isNaN(limit) ? 6 : limit;
  console.log({ category });

  const { data } = await heroApi.get<HeroesResponse>('/', {
    params: {
      limit,
      offset: (page - 1) * limit,
      category,
    },
  });

  const heroes = data.heroes.map((hero) => ({
    ...hero,
    image: `${BASE_URL}/images/${hero.image}`,
  }));

  return {
    ...data,
    heroes,
  };
};
