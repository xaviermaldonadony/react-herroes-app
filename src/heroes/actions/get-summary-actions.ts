import { heroApi } from '../api/hero.api';
import type { SummaryInformationResponse } from '../types/summary-information.response';

export const getSummaryAcion = async () => {
  const { data } = await heroApi.get<SummaryInformationResponse>('/summary');

  return data;
};
