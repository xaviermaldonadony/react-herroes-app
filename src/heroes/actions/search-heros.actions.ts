import { heroApi } from '../api/hero.api';
import type { Hero } from '../types/hero.interface';

interface Options {
  name?: string;
  team?: string;
  category?: string;
  universe?: string;
  status?: string;
  strength?: string;
}

const BASE_URL = import.meta.env.VITE_API_URL;

export const searchHeroesAction = async (options: Options = {}) => {
  const allValuesAreEmpty = Object.values(options).every((value) => !value);

  if (allValuesAreEmpty) {
    return [];
  }

  const { data } = await heroApi.get<Hero[]>('/search', {
    params: { ...options },
  });

  const heroes = data.map((hero) => ({
    ...hero,
    image: `${BASE_URL}/images/${hero.image}`,
  }));

  return heroes;
};
