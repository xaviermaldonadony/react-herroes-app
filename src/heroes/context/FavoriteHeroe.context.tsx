import type { Hero } from '../types/hero.interface';
import {
  createContext,
  useEffect,
  useState,
  type PropsWithChildren,
} from 'react';

interface FavoriteHeroContext {
  // state
  favorites: Hero[];
  favoriteCount: number;

  //   methods
  isFavorite: (hero: Hero) => boolean;
  toggleFavorites: (hero: Hero) => void;
}
// eslint-disable-next-line react-refresh/only-export-components
export const FavoriteHeroContext = createContext({} as FavoriteHeroContext);

const getFavoritesFromLocalStorage = (): Hero[] => {
  const favorites = localStorage.getItem('favorites');
  return favorites ? JSON.parse(favorites) : [];
};

export const FavoriteHeroeProvider = ({ children }: PropsWithChildren) => {
  const [favorites, setFavorites] = useState<Hero[]>(
    getFavoritesFromLocalStorage()
  );

  const toggleFavorites = (hero: Hero) => {
    const heroExist = favorites.find((h) => h.id === hero.id);

    if (heroExist) {
      setFavorites(favorites.filter((h) => h.id !== hero.id));
      return;
    }

    setFavorites([...favorites, hero]);
  };

  const isFavorite = (hero: Hero) => favorites.some((h) => h.id === hero.id);

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }, [favorites]);

  return (
    <FavoriteHeroContext
      value={{
        favoriteCount: favorites.length,
        favorites,

        // methods
        isFavorite,
        toggleFavorites,
      }}
    >
      {children}
    </FavoriteHeroContext>
  );
};
