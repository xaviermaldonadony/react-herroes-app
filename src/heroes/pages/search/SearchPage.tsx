import { useQuery } from '@tanstack/react-query';
import { useSearchParams } from 'react-router';

import { CustomJumbotron } from '@/components/custom/CustomJumbotron';
import { HeroStats } from '@/heroes/components/HeroStats';
import { SearchControls } from './ui/SearchControls';
import { CustomBreadcrumbs } from '@/components/custom/CustomBreadcrumbs';
import { HeroGrid } from '@/heroes/components/HeroGrid';
import { searchHeroesAction } from '@/heroes/actions/search-heros.actions';

export const SearchPage = () => {
  const [searchParams] = useSearchParams();
  const name = searchParams.get('name') ?? undefined;
  const strength = searchParams.get('strength') ?? undefined;

  const { data: heroes = [] } = useQuery({
    queryKey: ['searchHeroes', { name, strength }],
    queryFn: () => searchHeroesAction({ name, strength }),
    retry: 100 * 60 * 5,
  });

  return (
    <>
      <CustomJumbotron
        title={'Search for Super Heroe'}
        description='
          Discover, explore, and manage your favorite superheroes and villains'
      />
      <CustomBreadcrumbs
        currentPage='Search Heros'
        // breadcrumbs={[
        //   { label: 'Home', to: '/' },
        //   { label: 'Home2', to: '/' },
        //   { label: 'Home3', to: '/' },
        // ]}
      />
      {/* Stats Dashboard */}
      <HeroStats />
      {/* Filter and Search */}
      <SearchControls />

      <HeroGrid heroes={heroes} />
    </>
  );
};

export default SearchPage;
