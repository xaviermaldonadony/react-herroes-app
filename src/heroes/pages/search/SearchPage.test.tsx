import { FavoriteHeroeProvider } from '@/heroes/context/FavoriteHeroe.context';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { describe, expect, test, vi, beforeEach } from 'vitest';
import SearchPage from './SearchPage';
import { searchHeroesAction } from '@/heroes/actions/search-heros.actions';
import type { Hero } from '@/heroes/types/hero.interface';

vi.mock('@/heroes/components/HeroGrid', () => ({
  HeroGrid: ({ heroes }: { heroes: Hero[] }) => (
    <div data-testid='hero-grid'>
      {heroes.map((hero) => (
        <div key={hero.id}>{hero.name}</div>
      ))}
    </div>
  ),
}));

vi.mock('@/components/custom/CustomJumbotron', () => ({
  CustomJumbotron: () => <div data-testid='custom-jumbotron'></div>,
}));

vi.mock('@/heroes/actions/search-heros.actions');

vi.mock('./ui/SearchControls', () => ({
  SearchControls: () => <div data-testid='search-controls'></div>,
}));

const mockSearchHeroesAction = vi.mocked(searchHeroesAction);

const queryClient = new QueryClient();

const renderSearchPage = (initialEntries: string[] = ['/']) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <FavoriteHeroeProvider>
        <QueryClientProvider client={queryClient}>
          <SearchPage />
        </QueryClientProvider>
      </FavoriteHeroeProvider>
    </MemoryRouter>
  );
};

describe('SearchPage', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });
  test('Should render SearchPage with default values', () => {
    const { container } = renderSearchPage();

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: undefined,
      strength: undefined,
    });
    expect(container).toMatchSnapshot();
  });

  test('Should call earch action with name parameter', () => {
    const { container } = renderSearchPage(['/search?name=superman']);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: 'superman',
      strength: undefined,
    });

    expect(container).toMatchSnapshot();
  });

  test('Should call earch action with strength parameter', () => {
    const { container } = renderSearchPage([
      '/search?name=superman&strength=6',
    ]);

    expect(mockSearchHeroesAction).toHaveBeenCalledWith({
      name: 'superman',
      strength: '6',
    });

    expect(container).toMatchSnapshot();
  });

  test('Should render hero grid with search results', async () => {
    const mockHeroes = [
      { id: '1', name: 'Clark Kent' } as unknown as Hero,
      { id: '2', name: 'Bruce Wayne' } as unknown as Hero,
    ];

    mockSearchHeroesAction.mockResolvedValue(mockHeroes);

    renderSearchPage();

    await waitFor(() => {
      expect(screen.getByText('Clark Kent')).toBeDefined();
      expect(screen.getByText('Bruce Wayne')).toBeDefined();
    });
  });
});
