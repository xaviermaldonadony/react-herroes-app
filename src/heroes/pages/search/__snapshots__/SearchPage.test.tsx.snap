// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage > Should call earch action with name parameter 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors text-indigo-800 font-semibold"
          data-slot="breadcrumb-link"
        >
          Search Heros
        </a>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > Should call earch action with strength parameter 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors text-indigo-800 font-semibold"
          data-slot="breadcrumb-link"
        >
          Search Heros
        </a>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;

exports[`SearchPage > Should render SearchPage with default values 1`] = `
<div>
  <div
    data-testid="custom-jumbotron"
  />
  <nav
    aria-label="breadcrumb"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors text-indigo-800 font-semibold"
          data-slot="breadcrumb-link"
        >
          Search Heros
        </a>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="hero-grid"
  />
</div>
`;
